### PR TITLE
Add option to inflate unknown space

### PR DESCRIPTION
**Pull request by shrijitsingh99**
Signed-off-by: Shrijit Singh <shrijitsingh99@gmail.com>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1628 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | turtlebot simulation |

---

## Description of contribution in a few bullet points

* Added option to inflate unknown space
* Made some small optimizations and changes to make code cleaner
## Description of documentation updates required from your changes

* Added new parameter `inflate_around_unknown`, so need to add that to default configs and documentation page

### CODE REVIEW

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 1]
```diff
@@ -288,7 +289,7 @@ InflationLayer::enqueue(
     }
 
     // push the cell data onto the inflation list and mark
-    inflation_cells_[distance].push_back(CellData(index, mx, my, src_x, src_y));
+    inflation_cells_[distance].emplace_back(CellData(index, mx, my, src_x, src_y));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:292]: You can remove the `CellData` in this case if you're going to emplace_back.

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 2]
```diff
@@ -79,11 +80,13 @@ InflationLayer::onInitialize()
   declareParameter("inflation_radius", rclcpp::ParameterValue(0.55));
   declareParameter("cost_scaling_factor", rclcpp::ParameterValue(10.0));
   declareParameter("inflate_unknown", rclcpp::ParameterValue(false));
+  declareParameter("inflate_around_unknown", rclcpp::ParameterValue(false));
 
   node_->get_parameter(name_ + "." + "enabled", enabled_);
   node_->get_parameter(name_ + "." + "inflation_radius", inflation_radius_);
   node_->get_parameter(name_ + "." + "cost_scaling_factor", cost_scaling_factor_);
   node_->get_parameter(name_ + "." + "inflate_unknown", inflate_unknown_);
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:88]: What's this parameter? I haven't see this before: `inflate_unknown`. Did you test to see what this does / if that actually resolves the issue? 

I did not know myself about this parameter. It looks like we ported this from ROS1 but there's never been documentation about it in ROS1 or ROS2. This may be a hidden feature that actually fixes it.

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation/pull/839

This may also be good to include if you're interested in the inflation layer. Porting this to ROS2 would speed things up quite a bit.

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 2 above)
Comment (shrijitsingh99) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:86]: If set to true it allows the inflated area to extend into unknown space too else it gets cut of at unknown areas.

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:86]: got it, so its correct to say that do 2 different things.

**Comment by shrijitsingh99** [comment]
> I don't understand a bunch of these changes. Please explain. They don't see to follow the speed up PR or relate to inflating unknown space.

Since I was already working on the speedup you mentioned, I added some more optimizations and in general just cleaned up some stuff to be inline with modern practices:
* Convert `cached_costs_` and `cached_distances_` to vectors, they allow for automatic memory management, no need for `new` and `delete`.
* Convert `cached_costs_` and `cached_distances_` to 1D vectors, to allow better locality of reference.
* Introduced explicit static casts in-place of implicit and C-Styled casts and removed unnecessary casts.
* Added `override` keyword to mark implicit overriding of virtual functions.
* Fixed a issue mentioned in the code comments regarding `-numeric_limit<double>::max()`

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 2 above)
Comment (shrijitsingh99) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:86]: yeah, exactly

File: nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp [hunk 3]
```diff
@@ -108,21 +102,21 @@ class InflationLayer : public Layer
   inline unsigned char computeCost(double distance) const
   {
     unsigned char cost = 0;
+    double euclidean_distance = distance * resolution_;
     if (distance == 0) {
       cost = LETHAL_OBSTACLE;
-    } else if (distance * resolution_ <= inscribed_radius_) {
+    } else if (euclidean_distance <= inscribed_radius_) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:108]: Please revert this, you do an extra copy in that change + the variable isn't later re-used.

File: nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp [hunk 4]
```diff
@@ -133,11 +127,13 @@ class InflationLayer : public Layer
...
    */
-  inline double distanceLookup(int mx, int my, int src_x, int src_y)
+  inline double distanceLookup(
+    unsigned int mx, unsigned int my, unsigned int src_x,
+    unsigned int src_y)
   {
-    unsigned int dx = abs(mx - src_x);
-    unsigned int dy = abs(my - src_y);
-    return cached_distances_[dx][dy];
+    unsigned int dx = (mx > src_x) ? mx - src_x : src_x - mx;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:134]: This is nearly identical to the implementation of abs (https://code.woboq.org/userspace/glibc/stdlib/abs.c.html), please revert for readability.

File: nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp [hunk 5]
```diff
@@ -148,16 +144,16 @@ class InflationLayer : public Layer
...
    */
-  inline unsigned char costLookup(int mx, int my, int src_x, int src_y)
+  inline unsigned char costLookup(
+    unsigned int mx, unsigned int my, unsigned int src_x,
+    unsigned int src_y)
   {
-    unsigned int dx = abs(mx - src_x);
-    unsigned int dy = abs(my - src_y);
-    return cached_costs_[dx][dy];
+    unsigned int dx = (mx > src_x) ? mx - src_x : src_x - mx;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:151]: This is nearly identical to the implementation of abs (https://code.woboq.org/userspace/glibc/stdlib/abs.c.html), please revert for readability.

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 6]
```diff
@@ -61,14 +60,14 @@ InflationLayer::InflationLayer()
...
   cell_inflation_radius_(0),
   cached_cell_inflation_radius_(0),
-  cached_costs_(nullptr),
-  cached_distances_(nullptr),
-  last_min_x_(-std::numeric_limits<float>::max()),
-  last_min_y_(-std::numeric_limits<float>::max()),
-  last_max_x_(std::numeric_limits<float>::max()),
-  last_max_y_(std::numeric_limits<float>::max())
+  cache_length_(0),
+  last_min_x_(std::numeric_limits<double>::lowest()),
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:67]: lowest appears to just be -max  ( https://en.cppreference.com/w/cpp/types/numeric_limits). Not sure what issue you were referring to that asked for that to be changed. Can you clarify / link?

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 7]
```diff
@@ -112,13 +115,11 @@ InflationLayer::updateBounds(
...
     last_max_y_ = *max_y;
-    // For some reason when I make these -<double>::max() it does not
-    // work with Costmap2D::worldToMapEnforceBounds(), so I'm using
-    // -<float>::max() instead.
-    *min_x = -std::numeric_limits<float>::max();
-    *min_y = -std::numeric_limits<float>::max();
-    *max_x = std::numeric_limits<float>::max();
-    *max_y = std::numeric_limits<float>::max();
+
+    *min_x = std::numeric_limits<double>::lowest();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:119]: lowest appears to just be -max ( https://en.cppreference.com/w/cpp/types/numeric_limits). Not sure what issue you were referring to that asked for that to be changed. Can you clarify / link?

Comment (shrijitsingh99) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:119]: https://github.com/ros-planning/navigation2/blob/ecc42fb3879d470590b9d9ef893c7565cee3b69b/nav2_costmap_2d/plugins/inflation_layer.cpp#L115-L117

File: nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp (hunk 4 above)
Comment (shrijitsingh99) [reply nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:134]: I did this as initially `distanceLookup(int mx, int my, int src_x, int src_y)` was this but where ever it was called its arguments were `unsigned int`:
https://github.com/ros-planning/navigation2/blob/d21cf3366b06e3a63014e411708c2cbae2ae3647/nav2_costmap_2d/plugins/inflation_layer.cpp#L235

So it had to make 2 casts one from unsigned int to int then back to unsigned int when assigning to `dx` and `dy`.
This way there are no casts.

**Comment by shrijitsingh99** [comment]
> There are a bunch of missing changes that Jorge implemented. At the bare minimum you're missing
> 
> https://github.com/ros-planning/navigation2/blob/d21cf3366b06e3a63014e411708c2cbae2ae3647/nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp#L171
> the change between a map to a vector, which I think accounts for the vast majority of the speed up.
> 
I have that port ready, haven't pushed it currently.

> I think you should simplify to a direct port before trying to do anything else. Some of the changes you made actually will make the code a little slower.
I mostly made changes related to recommended C+= syntax so shouldn't negatively affect performance.

Haven't benchmarked the `cached_costs_` and `cached_distances_` but should be slightly faster due to locality, worst case there is no performance impact. 

Might have missed some stuff causing it to be slower. If you think all the changes are unneeded, will revert back.

File: nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:134]: Ah ok, I can live with that

**Comment by SteveMacenski** [comment]
I would make all the changes he made. Jorge's a smart guy and if those changes were robustly checked from months of running on Magazino's robots, I'd go with that and take his changes at face value.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 it fails some linters and the unit tests for the inflation layer. Please make sure to run these tests locally + run the stack with these change to debug. It looks like there's a real bug in there.

**Comment by shrijitsingh99** [comment]
Yeah sorry about that. Have made separate issue for a bug in the tests. Lets get that merged first before proceeding with this. Opened a separate PR since I don't think it is a good idea modifying the test as well as the module together can lead to bugs creeping in.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 are you running on ROS2 master source build or Eloquent binaries?

You'd see crustify appearing differently if you did Eloquent, there was a master / Foxy update. Its basically just:
```
auto yo = some_function_long_name(first_thing,
  second_thing);
```

goes to 

```
auto yo = some_function_long_name(
  first_thing, second_thing);
```

Any line extensions start arguments on a new line.

**Comment by shrijitsingh99** [comment]
> @shrijitsingh99 are you running on ROS2 master source build or Eloquent binaries?

 Eloquent binaries

> You'd see crustify appearing differently if you did Eloquent, there was a master / Foxy update.

Yup, pretty much sums the issue I was facing. Will switch to source build.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 you should rebase to get the new test

**Comment by SteveMacenski** [comment]
@shrijitsingh99 can you verify you see a CPU drop? I want to make sure we got what we came for with this.

**Comment by SteveMacenski** [comment]
@shivaang12 please review

**Comment by SteveMacenski** [comment]
Inflation layer is actually one of the best covered files in the repo https://codecov.io/gh/ros-planning/navigation2/src/master/nav2_costmap_2d/plugins/inflation_layer.cpp

Though I think that is mostly from integration testing that wouldn't actually come up with any issues unless it was _really_ messed up since the robot an still navigate as long as its not in the middle of LETHALs. 

@shrijitsingh99 that is a good point though. One of the very few uncovered lines is this PR adds https://codecov.io/gh/ros-planning/navigation2/src/master/nav2_costmap_2d/plugins/inflation_layer.cpp#L239 it would be helpful to add a test for checking with this param on / off that it inflates or doesn't inflate unknown space. 

An easy test would be creating a 10x10 costmap of FREE and in the middle a single unknown cell. Check when off that its neighbors are still FREE, when on, checking that they're > FREE.

**Comment by shrijitsingh99** [comment]
Great, will add some tests for `inflate_unknown_` and `inflate_around_unknown_`.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 any update on this? We should try to square up existing things before adding new things. 5 unmerged PRs isn't as good as 3 merged ones :wink:

**Comment by shrijitsingh99** [comment]
> @shrijitsingh99 any update on this? We should try to square up existing things before adding new things. 5 unmerged PRs isn't as good as 3 merged ones 😉

I have started work on it should be done by tomorrow morning.

> You will also need to rebase this on master, a bunch of emergency updates went in over the weekend.

Will do that 👍

File: nav2_costmap_2d/test/integration/inflation_tests.cpp [hunk 8]
```diff
@@ -137,7 +138,8 @@ void TestNode::validatePointInflation(
         double dist = std::hypot(dx, dy);
 
         unsigned char expected_cost = ilayer->computeCost(dist);
-        ASSERT_TRUE(costmap->getCost(cell.x_, cell.y_) >= expected_cost);
+//        ASSERT_TRUE(costmap->getCost(cell.x_, cell.y_) >= expected_cost)
```
Comment (SteveMacenski) [inline nav2_costmap_2d/test/integration/inflation_tests.cpp:141]: Why is this commented out?

Comment (shrijitsingh99) [reply nav2_costmap_2d/test/integration/inflation_tests.cpp]: Was testing out something,, mistakenly commented out.
Fixed.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 did you mess with build settings or setup a CI on your own fork? That build error is not from our stuff and when I click on it it shows as part of your account/fork

**Comment by shrijitsingh99** [comment]
> @shrijitsingh99 did you mess with build settings or setup a CI on your own fork? That build error is not from our stuff and when I click on it it shows as part of your account/fork

Yeah, I was trying to set up the I on my own fork for testing purposes.

I didn't know it will switch to my personal CI for PR's too. How should I do go about resolving this?

**Comment by SteveMacenski** [comment]
> I didn't know it will switch to my personal CI for PR's too. How should I do go about resolving this?

A PR can be thought of as "I have my own version of this [thing] and I want to generate a diff and apply that patch to your [thing]". When the MR is opened, its going to look at your end of things to decide what to trigger/do so it'll naturally run your configs. In fact, you can open a pull request from a completely unrelated repository into this one if you liked and applied some random patch. The GitHub interface doesn't make that easy, but Git will fully allow for it.

tl;dr, change the branch that run your CI stuff against to something other than `master` or `*-devel`

**Comment by shrijitsingh99** [comment]
> When the MR is opened, its going to look at your end of things to decide what to trigger/do so it'll naturally run your configs. 

I generally have been using Azure Pipelines, it allows to me run my own CI on my fork, and on the PR it runs the upstream repo's CI so I assumed here there was a similar thing.

Is this CircleCI specific or am I missing something?

PS: Have unlinked my own CI.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 don't bother, @AlexeyMerzlyakov should, he's been thinking about this problem. Merging this since that's not part of this PR.

**Comment by SteveMacenski** [comment]
Actually,  @shrijitsingh99 can you do one thing on this, can you branch off this branch, rebase on master as of right now, and then resubmit a PR with some trivial change? (add a comment, whatever) 

I want to see if this is from some state issue. It can be a draft PR
